parlai/tasks/wizard_of_internet/agents.py

```diff
@@ -0,0 +1,564 @@
+        teacher_action: Message,
+        labels: Optional[Tuple[str]],
+        model_response: Message,
+    ) -> None:
+        if (
+            teacher_action[CONST.SELECTED_SENTENCES][0]
+            == CONST.NO_SELECTED_SENTENCES_TOKEN
+        ):
+            # Has NOT selected knowledge
+            return
+
+        resp = model_response['text']
+        self.metrics.add(
+            'knowledge_f1_docs',
+            F1Metric.compute(resp, [' '.join(teacher_action[CONST.SELECTED_DOCS])]),
+        )
+        self.metrics.add(
+            'knowledge_f1_max_docs', F1Metric.compute(resp, CONST.SELECTED_DOCS)
+        )
+        self.metrics.add(
+            'knowledge_f1_sentences',
+            F1Metric.compute(
+                resp, [' '.join(teacher_action[CONST.SELECTED_SENTENCES])]
+            ),
+        )
+        self.metrics.add(
+            'knowledge_f1_max_sentences',
+            F1Metric.compute(resp, CONST.SELECTED_SENTENCES),
+        )
+
+    def _teacher_action_type(self) -> str:
+        return CONST.ACTION_WIZARD_TO_APPRENTICE
+
+    def additional_message_content(self, parlai_message: Message, action: Dict):
+        for item_key in (
+            CONST.RETRIEVED_DOCS,
+            CONST.RETRIEVED_DOCS_URLS,
+            CONST.SELECTED_DOCS,
+            CONST.SELECTED_SENTENCES,
+            CONST.SEARCH_QUERY,
+        ):
+            parlai_message[item_key] = action[item_key]
+
+    def teacher_setup_data(self, datafile) -> Message:
+        for message, episode_started in super().teacher_setup_data(datafile):
+            if self.prepend_gold_knowledge:
+                text = message[CONST.MESSAGE_TEXT]
+                gold_knowledge = self.gold_knowledge_delimiter.join(
+                    message[CONST.SELECTED_SENTENCES]
+                )
+                message.force_set(
+                    CONST.MESSAGE_TEXT,
+                    (
+                        f'{CONST.KNOWLEDGE_TOKEN} {gold_knowledge} {CONST.END_KNOWLEDGE_TOKEN}'
+                        f' {self.gold_knowledge_delimiter} {text}'
+                    ),
+                )
+            yield message, episode_started
+
+
+class WizardDialogGoldKnowledgeTeacher(WizardDialogTeacher):
+    @classmethod
+    def add_cmdline_args(cls, parser: ParlaiParser, partial_opt=None) -> ParlaiParser:
+        super().add_cmdline_args(parser, partial_opt)
+        # This teacher is only allowed with non-episodic examples
+        parser.set_params(episodic_examples=False)
+        return parser
+
+
+class DefaultTeacher(WizardDialogTeacher):
+    pass
+
+
+###############################################################
+#                                                             #
+# Search and Knowledge Teachers                               #
+#                                                             #
+###############################################################
+
+
+class BaseSQKnowledgeTeacher(WizardOfInternetBaseTeacher):
```

Comment:
maybe note that it's abstract/shouldnt be instantiated directly? @mojtaba-komeili